### PR TITLE
[primTorch] Add ref for `huber_loss` and error inputs

### PR DESCRIPTION
**Pull request by nkaretnikov**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85041

### CODE REVIEW

File: torch/_refs/nn/functional/__init__.py
Comment (Lezcano) [inline torch/_refs/nn/functional/__init__.py:445]: Obs. This is fine for real numbers, as taking `abs` means we just multipyl by the sign. For complex numbers, it'd be better to do `diff = input - target` and then `diff * dif.conj()`. Now, this loss does not support complex numbers, so that's completely fine .

**Review by lezcano — APPROVED** [review]
Cool!

File: torch/_refs/nn/functional/__init__.py
Comment (ngimel) [inline torch/_refs/nn/functional/__init__.py:433]: the only difference is in reduction argument? I think we can make reference accept both string and integer reduction argument.

File: torch/_refs/nn/functional/__init__.py
Comment (ngimel) [inline torch/_refs/nn/functional/__init__.py:430]: do you need `out` wrapper here?

File: torch/_refs/nn/functional/__init__.py
Comment (ngimel) [reply torch/_refs/nn/functional/__init__.py:445]: `complex_tensor.abs()` would do the right thing, no?

Comment (Lezcano) [reply torch/_refs/nn/functional/__init__.py:445]: For complex tensors, we would have an expression like `diff.abs()**2`, but you can save the `sqrt`  within `abs()` by writing it as `diff.abs() ** 2 == diff * diff.conj()`. But again, this loss does not support complex numbers in core, so that's fine.

File: torch/_refs/nn/functional/__init__.py
Comment (mruberry) [reply torch/_refs/nn/functional/__init__.py:430]: Interestingly `nn.functional.huber_loss` doesn't support an `out` parameter, which is a UX consistency bug:

https://github.com/pytorch/pytorch/blob/d52452b3d1ff14a87e1d79c8d8bd67f0f074e6e6/torch/nn/functional.py#L3199

We should probably scrub and correct these UX consistency issues together. I made a note to follow-up.

File: torch/_refs/nn/functional/__init__.py
Comment (mruberry) [reply torch/_refs/nn/functional/__init__.py:445]: We could add a comment about future possible complex number support

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:7019]: Nice error checks

File: torch/_refs/nn/functional/__init__.py
Comment (ngimel) [reply torch/_refs/nn/functional/__init__.py:433]: How about doing it the other way around, and have torch reference map a str argument to an int, and the call _decomp aten decomposition?

Comment (mruberry) [reply torch/_refs/nn/functional/__init__.py:433]: > How about doing it the other way around, and have torch reference map a str argument to an int, and the call _decomp aten decomposition?

It doesn't really matter which way it would be done today

File: torch/_refs/nn/functional/__init__.py
Comment (nkaretnikov) [reply torch/_refs/nn/functional/__init__.py:430]: no, i think. the yaml file has the out overload, but the opinfo and the function signature say that `out` is not supported

Comment (Lezcano) [reply torch/_refs/nn/functional/__init__.py:430]: Note that this is an issue with many of the loss functions really: the function has an out variant on ATen, but it's not exposed on the `functional` API.

Comment (nkaretnikov) [reply torch/_refs/nn/functional/__init__.py:430]: > We should probably scrub and correct these UX consistency issues together. I made a note to follow-up.

note: i have a wip local branch that allows testing for `out` support consistency between regular pytorch ops and primtorch refs. it needs some clean up, but i'm planning to submit once the ref alias checking stack is merged

Comment (Lezcano) [reply torch/_refs/nn/functional/__init__.py:430]: Note that it's fine if a reference has the `out=` behaviour but the original operation doesn't. Also, we already have a number of skips/errors for `test_out` in core, so I'd expect that, if anything, the references are more correct than the functions in core. It may be a bit tricky to write a test that's not too flaky and that test all these issues.

**Comment by nkaretnikov** [comment]
@mruberry @ngimel addressed feedback, ci is green, ptal

**Review by mruberry — APPROVED** [review]
Cool!

**Comment by nkaretnikov** [comment]
@ngimel a gentle reminder that this one is waiting for review, PTAL

**Comment by ngimel** [comment]
This is approved, you can land it.

**Pull request closed without merging**
